samcli/commands/build/build_context.py

```diff
@@ -307,6 +311,105 @@ def run(self):
+                        "Sourcemap set without --enable-source-maps, adding"
+                        " --enable-source-maps to function's NODE_OPTIONS"
+                    )
+
+                    resource.setdefault("Properties", {})
+                    resource["Properties"].setdefault("Environment", {})
+                    resource["Properties"]["Environment"].setdefault("Variables", {})
+                    existing_options = resource["Properties"]["Environment"]["Variables"].setdefault("NODE_OPTIONS", "")
+
+                    # make sure the NODE_OPTIONS is a string
+                    if not isinstance(existing_options, str):
+                        invalid_node_option = True
+                    else:
+                        resource["Properties"]["Environment"]["Variables"]["NODE_OPTIONS"] = " ".join(
+                            [existing_options, "--enable-source-maps"]
+                        )
+
+                    using_source_maps = True
+
+                # check if --enable-source-map is provided and append Sourcemap: true if it is not set
+                if source_map is None and node_option_set:
+                    LOG.debug(
+                        "--enable-source-maps set without Sourcemap, adding Sourcemap to Metadata BuildProperties"
+                    )
+
+                    resource.setdefault("Metadata", {})
+                    resource["Metadata"].setdefault("BuildProperties", {})
+                    resource["Metadata"]["BuildProperties"]["Sourcemap"] = True
+
+                    using_source_maps = True
+
+        if using_source_maps:
+            self._warn_using_source_maps()
+
+        if invalid_node_option:
+            self._warn_invalid_node_options()
+
+    @staticmethod
+    def _is_node_option_set(resource: Dict) -> bool:
+        """
+        Checks if the template has NODE_OPTIONS --enable-source-maps set
+
+        Parameters
+        ----------
+        resource : Dict
+            The resource dictionary to lookup if --enable-source-maps is set
+
+        Returns
+        -------
+        bool
+            True if --enable-source-maps is set, otherwise false
+        """
+        try:
+            node_options = resource["Properties"]["Environment"]["Variables"]["NODE_OPTIONS"]
+
+            return "--enable-source-maps" in node_options.split()
+        except (KeyError, AttributeError):
+            return False
+
+    @staticmethod
+    def _warn_invalid_node_options():
+        click.secho(
+            "\nNODE_OPTIONS is not a string! As a result, the NODE_OPTIONS environment variable will "
+            "not be set correctly, please make sure it is a string. "
+            "Visit https://nodejs.org/api/cli.html#node_optionsoptions for more details.\n",
+            fg="yellow",
+        )
+
+    @staticmethod
+    def _warn_using_source_maps():
+        click.secho(
+            "\nYou are using source maps, note that this comes with a performance hit!"
```

Comment:
If they try to remove just the env var it will still be set next time if the source maps are still set in the metadata, right? Should we clarify that it needs to be removed from both spots?

Reply:
Nice catch, I'll update it to reflect this change